### ISSUE TITLE
Snapshot copying and `StaticsContainer`/`KineticsContainer`

### ISSUE THREAD

**Issue by dwhswenson** [openpathsampling/openpathsampling#580]
In working on #579, I've come across what I think is a bug in the behavior of `snapshot.copy()` when using OpenMM snapshots.

Short version: based on our (long!) discussions in #429 and #444, I thought that `snapshot.copy()` was supposed to essentially be a deep copy for data (shallow for other things, like topology/engine). Importantly, it is a deep copy for numpy data, like in the toy engine's velocities and coordinates. However, it doesn't seem to work this way for the OpenMM snapshot, where the coordinates and velocities are wrapped in the `statics` or `kinetics`.

Quick example:

```python
import numpy as np
import openmmtools as omt
import openpathsampling.engines.toy as toy_engine
import openpathsampling.engines.openmm as omm_engine
import simtk.unit as u

# test a toy snapshot
init_coord = np.array([1.0, 2.0])
init_vel = np.array([3.0, 4.0])
toy_snap = toy_engine.Snapshot(coordinates=init_coord, velocities=init_vel)
toy_copy = toy_snap.copy()
toy_copy.velocities[1] = 1
print toy_snap.velocities, toy_copy.velocities
# [ 3.  4.] [ 3.  1.]

# test an openmm snapshot
sys = omt.testsystems.AlanineDipeptideVacuum()
omm_snap = omm_engine.snapshot_from_testsystem(sys)
# initial velocity
print omm_snap.velocities[0]
# [ 0.  0.  0.] nm/ps

omm_copy = omm_snap.copy()
omm_copy.velocities[0] = [1.0, 2.0, 3.0] * u.nanometer / u.picosecond
print omm_snap.velocities[0], omm_copy.velocities[0]
# [ 1.  2.  3.] nm/ps [ 1.  2.  3.] nm/ps

print hex(id(omm_snap.kinetics)), hex(id(omm_copy.kinetics))
# 0x11104e510 0x11104e510
```

In general, we need a solution for this to do 2-way shooting. It needs to work with either the toy engine or OpenMM, so it can't *require* `statics` or `kinetics` objects. The things to balance here are:

* we don't want to always make new `statics`/`kinetics` when we make a copy of the snapshot
* we want the option to tell the snapshot to make a deep copy of container attributes like `statics` or `kinetics`
* we want this to be independent of the details of the snapshot: it shouldn't care whether we have container objects or direct numpy arrays (i.e., OpenMM vs. Toy snapshots)

The best idea I've come up with looks something like this: Add another type of `copy`, something like `snapshot.copy_with_deepcopied_attributes(['velocities'])`, which would make a copy of the given snapshot that guarantees the `velocities` attribute is deep-copied. For a toy snapshot, this wouldn't be exactly the same as the normal `copy()`, although it might differ if the attribute list included, e.g., `engine` (maybe some attributes could forbid themselves from deep copy?). However, for OpenMM, this would need to be smart enough to recognize that `'velocities'` is defined as a part of `kinetics`, and therefore it needs to `deepcopy` the `kinetics` object.

The point is that immediately after making such a copy, it is safe to change these attributes (this copy can not have already been saved). So if you're writing a `SnapshotModifier` that changes the velocities, you make the new version of the snapshot as in the previous paragraph. If you'll modify both coordinates and velocities, you do `snapshot.copy_with_deepcopied_attributes(['coordinates', 'velocities'])`.

Now, this is an idea for how this could work, but I'm much less certain on the implementation. @jhprinz, all of this revolves around the dynamic snapshot creation, as well as the `kinetics` and `statics` stuff -- all of which is your turf. How hard would it be to implement this sort of thing?

**Comment by jhprinz** [comment]
Indeed. In short: `KineticContainer` and `StaticContainer` is an OPS object and hence it will not be deep.copied. You can either use the Snapshot, that does not use these containers oder you create a copy _by hand_ or you add a `.copy` method to the class definition. I think the last would be the cleanest choice. The features only provide fallbacks, when you do not provide your own function. And in this case I assume that needs to be fixed by hand. 

I need to think about a more long-term approacha little. 

Can you also explain where this `copy` happens? Is it in a mover or inside an engine? I think this could be an engine function and hence be the same for toy and openmm. If openmm uses `statics` it knows what to do. 

If you want a special function like you describe this needs to be implemented manually for all snapshots. I am not sure if we can inspect the way statics/kinetics or something much more complicated is build and what needs to happen to duplicate a certain attribute. 

It is like saying that the openmm snapshots have `velocities` which are accessed by `kinetics.velocities` the same way I would write a function `copy_with_new_velocities()` which has a default and for openmm snapshots needs to be overridden.

I will think about this.

**Comment by dwhswenson** [comment]
Thanks for the quick response. Right now my master's student is manually creating snapshots with different velocities for the thing we need immediately. We can make it work, but it is an ugly hack. And the fact that we need it so soon means that this will be an obvious lack to users. `CommittorSimulator` does not work with OpenMM. I also can't complete my checks of #579 until we come up with a fix for this.

> You can either use the Snapshot, that does not use these containers oder you create a copy *by hand* or you add a `.copy` method to the class definition. I think the last would be the cleanest choice.

Not sure I understand here. Do you mean the same as the `copy_with_new_velocities` you mention later? If not, do you mean that we should overwrite the OpenMM snapshot's `copy` function? Same for every single snapshot type? (That doesn't sound good.)

> Can you also explain where this `copy` happens? Is it in a mover or inside an engine?

`SnapshotModifier`. In this case, I'm completely randomizing velocities according to the Boltzmann distribution. In another case, I might rotate the direction of the velocity of certain atoms by some randomly-selected amount. In yet another case, I might swap velocities of a pair of atoms. Even more generally, I might do any sort of configurational Monte Carlo move (it doesn't have to be velocities that change).

This is essential for two-way shooting. The current work is to make the committor analysis work in OpenMM.

> If you want a special function like you describe this needs to be implemented manually for all snapshots. I am not sure if we can inspect the way statics/kinetics or something much more complicated is build and what needs to happen to duplicate a certain attribute.

Add a `provides` or `properties` list to each feature which lists any properties it includes? That makes inspection trivial. Since we already know the names of the attributes it adds, we know which things need to be deep-copied. Also add a `no_deep_copy` list which allows you to forbid certain things from deep-copy.

Note also that the energies will be a lie after this. They would need to be re-calculated or set to `None`. Maybe a custom `_deepcopy()` in the feature could be used to set them to either correct or `None` values?

> It is like saying that the openmm snapshots have `velocities` which are accessed by `kinetics.velocities` the same way I would write a function `copy_with_new_velocities()` which has a default and for openmm snapshots needs to be overridden.

This sounds to me like it would break everything about the automatic snapshot generation, at least if you use any container features. You need `copy_with_new_velocities()`, `copy_with_new_coordinates()`, `copy_with_new_box_vectors()`, and who knows what else.

**Comment by dwhswenson** [comment]
Perhaps a better idea: what about a function `snapshot.copy_with_replacement(self, feature1=None, feature2=None, ...)`? This would act like a generalization of the `copy_with_new_<feature>()` idea.

If `None`, the new snapshot copies the variable as in the current `snapshot.copy()`. If not `None`, you use the given value in place of the existing one.

In the case I'm working on, I would make the modifications to a copy of the velocities from the relevant snapshot. Then I would call:

```python
new_snap = old_snap.copy_with_replacement(velocities=new_velocities)
```

Again, this might require some additional tricks to get it to work with containers -- for now, maybe just override it for OpenMM? I think this could be dynamically generated for simple cases where the variables don't depend on each other (the way energies depend on coords/vels).

This seems pretty easy: a single function, and a `TypeError` given if you provide a nonsense argument for the thing you want to replace. In terms of what I'm trying to do, it's easy enough to get the `velocities`, make a (deep)copy of that `nparray`, make the modifications, and include that as an argument to `copy_with_replacement` (and it seems like a cleaner approach, too).

**Comment by jhprinz** [comment]
Another idea that might even be easier. it work under the following assumption:

Basically what we want is that `.copy` will really duplicate the statics and kinetics object, but no all other snapshot objects, like the engine. So the `statics` feature needs to know that if you copy, that its kinetics and statics objects need to be duplicated, too. This would be simple to implement. Not trivial but quickly doable.

Now, in your case you want even more flexibility. You want to be able to make a copy but selectively duplicate objects without even knowing about it. I mean the idea of having `statics` and `kinetics` was

1. to show off, that this can be done
2. to only change the velocities and reuse the coordinates

If you do not care to save the coordinates twice then you could use another `Snapshot` class that avoids statics and kinetics. Also note, that in case of the statics/kinetics `Snapshot` you need to use the `Snapshot.construct` method to generate a snapshot exclusively from coordinates and velocities. This function has somehow the same problem.

I guess the problem will be that `velocities` and `coordinates` are only properties but not stored values. So if you want to replace these then what!?! We need some logic to replace these, like a setter. So that, if you set velocities the internal kinetics object will be replaced.

But, what if you replace only box_vectors? These are coupled with coordinates and making a new object will need a new statics object. I think that is not a problem. The only issue (performance wise) is that if you change the box_vectors and coordinates, it will first create a new statics with new coordinates and then another new object with also box_vectors changed.

If that is not a problem, then I think the `copy_with_replacement` is very easy to implement. it will look like this

```py
def copy_with_replacement(self, **kwargs):
    cp = self.copy()  # this will copy all, but it is simple
    for key, value in kwargs.iteritems():
        setattr(cp, key, value)  # this triggers the setter and will replace the internal values
```

in case of toy snapshots it will simply replace the attributes. For the OpenMM attribute it will trigger the setter which do what it necessary.

Would that be acceptable?

**Comment by dwhswenson** [comment]
> Basically what we want is that `.copy` will really duplicate the statics and kinetics object, but no all other snapshot objects, like the engine. So the `statics` feature needs to know that if you copy, that its kinetics and statics objects need to be duplicated, too. This would be simple to implement. Not trivial but quickly doable.

I had assumed this approach was off the table because we didn't want to have different `statics` from before and after a 2-way shooting point. From a storage standpoint, this makes a tiny difference (~0.5%, I would guess), so I don't think it actually matters, but it might make some kinds of analysis easier.

> The only issue (performance wise) is that if you change the box_vectors and coordinates, it will first create a new statics with new coordinates and then another new object with also box_vectors changed.

This doesn't bother me at all. The frequency of doing this should be on the order of once per trajectory (so probably once per 100 MD frames or less) -- that extra cost won't be noticed.

> Would that be acceptable?

That definitely works -- and especially if it is as easy as that code. It also needs something to make the setters work, in the OpenMM case, but I don't think those will be too painful, right?

Thanks!

**Comment by dwhswenson** [comment]
One note: it's important to make sure that the `copy_with_replacement` gives the expected error on a bad `kwargs`. Correct behavior from example:

```python
def foo(bar):
    pass

foo(baz="qux")
```

```pytb
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-6ec7222115e9> in <module>()
----> 1 foo(baz="qux")

TypeError: foo() got an unexpected keyword argument 'baz'
```

This will probably involve catching some other error (`AttributeError`? even `KeyError`, depending on internals?) and re-raising it as a `TypeError`. Easy enough to do, but worth keeping in mind!

**Comment by dwhswenson** [comment]
Resolved by #581 (apparently "Addresses" isn't one of the [special keywords to close an issue from a PR](https://help.github.com/articles/closing-issues-via-commit-messages/))
